Add spark.dotnet.ignoreSparkPatchVersionCheck conf to ignore patch version in DotnetRunner

suhsteve opened the pull request:
Add a new spark conf, `spark.dotnet.ignoreSparkPatchVersionCheck`.  If set to true, then only check the major.minor `SPARK_VERSION` and ignore patch version.  If set to false, then check the `SPARK_VERSION` against the set of `supportedSparkVersions`

imback82 commented on src/scala/microsoft-spark-3-0/src/main/scala/org/apache/spark/deploy/dotnet/DotnetRunner.scala:
```diff
@@ -164,12 +167,18 @@ object DotnetRunner extends Logging {
-    if (!supportedSparkVersions(SPARK_VERSION)) {
+  private def validateSparkVersions(conf: SparkConf): Unit = {
+    val ignorePatchVersion = conf.get(DOTNET_IGNORE_SPARK_PATCH_VERSION_CHECK)
+    val majorMinorVersion = supportedSparkMajorMinorVersion.mkString("", ".", ".")
```
just store it as a string to avoid `mkString`?

suhsteve replied:
Sure.  Do you want me to hardcode it in this method or set it as a class field at the top of the class definition ?

imback82 replied:
Class field seems fine to me.

imback82 commented on src/scala/microsoft-spark-3-0/src/main/scala/org/apache/spark/deploy/dotnet/DotnetRunner.scala:
```diff
@@ -164,12 +167,18 @@ object DotnetRunner extends Logging {
-          s" Supported versions: ${supportedVersions}")
+        s"Unsupported spark version used: ${spark.SPARK_VERSION}. Normalized spark version used: $SPARK_VERSION." +
+          s" Supported versions: $supportedVersions")
     }
```
Can we add warning message that it's being bypassed?

imback82 reviewed (commented):
I assume you tested this manually? How about just put this in Utils, so that we can unit-test this?

suhsteve commented:
> I assume you tested this manually? How about just put this in Utils, so that we can unit-test this?

Yup, was manually tested.  I can move to Utils.

suhsteve commented:
> > I assume you tested this manually? How about just put this in Utils, so that we can unit-test this?
> 
> Yup, was manually tested. I can move to Utils.

Moved to Utils and added tests.

imback82 commented:
Can you check the test failures?

imback82 commented on src/scala/microsoft-spark-2-3/src/test/scala/org/apache/spark/util/dotnet/UtilsTest.scala:
```diff
@@ -0,0 +1,80 @@
+    val supportedSparkVersions = Set[String]("2.3.0", "2.3.1", "2.3.2", "2.3.3", "2.3.4")
+
+    assertThrows(
+      classOf[IllegalArgumentException],
```
Can you use `intercept` and check the message as well (since we have two different places where we throw exceptions)? https://github.com/microsoft/hyperspace/blob/053b6b1e00f2b6dd62889e40d417dff1939079fc/src/test/scala/com/microsoft/hyperspace/index/CreateIndexTest.scala#L76

imback82 replied:
Or we don't have `scalatest` set up?

imback82 replied:
If `scalatest` is not possible, can you grep the exception message and make sure it's thrown at a different place?

suhsteve replied:
We don't have scalatest set up, only junit.  But we can still check exception message with it.

imback82 commented on src/scala/microsoft-spark-2-3/src/main/scala/org/apache/spark/deploy/dotnet/DotnetRunner.scala:
```diff
@@ -43,7 +44,12 @@ object DotnetRunner extends Logging {
 
-    validateSparkVersions
+    DotnetUtils.validateSparkVersions(
+      Option(SparkEnv.get).map(_.conf).getOrElse(new SparkConf()),
```
question. this SparkEnv will be set for `spark-submit` and livy submission right?

suhsteve replied:
I know it works for `spark-submit`, not sure about livy submission.

imback82 replied:
Hmm, to me `SparkEnv.get` should always return None since it's set in during SparkContext initialization? I think the reason this works is because of
https://github.com/apache/spark/blob/06d40696dc27c39824666a70e7c673880959a681/core/src/main/scala/org/apache/spark/SparkConf.scala#L70

And
https://github.com/apache/spark/blob/06d40696dc27c39824666a70e7c673880959a681/core/src/main/scala/org/apache/spark/deploy/SparkApplication.scala#L47

Were you getting `Some` from `SparkEnv.get`?

suhsteve replied:
`SparkEnv.get` is `None`.  We can pass in a `new SparkConf` directly or just get the values from `sys.props` instead if you prefer.

imback82 replied:
If it's always None, I would remove it since it just adds confusion and document clearly. For example:
```scala
// Load defaults set in the system properties when this application is launched:
// https://github.com/apache/spark/blob/06d40696dc27c39824666a70e7c673880959a681/core/src/main/scala/org/apache/spark/deploy/SparkApplication.scala#L47
val sparkConf = new SparkConf(loadDefaults = true)
```

suhsteve replied:
I opted to go with the `sys.props` since [RRunner.scala](https://github.com/apache/spark/blob/0494dc90af48ce7da0625485a4dc6917a244d580/core/src/main/scala/org/apache/spark/deploy/RRunner.scala#L48) is also getting the conf this way.

imback82 replied:
OK, either one is fine as long as `SparkEnv.get` is removed.

imback82 reviewed (commented):
LGTM (few minor comments), thanks @suhsteve!

suhsteve commented on src/scala/microsoft-spark-2-3/src/main/scala/org/apache/spark/util/dotnet/Utils.scala:
```diff
@@ -191,6 +193,45 @@ object Utils extends Logging {
+   * @param conf Spark conf.
+   * @param sparkVersion The spark version.
+   * @param normalizedSparkVersion: The normalized spark version.
+   * @param sparkMajorMinorVersionPrefix The spark major and minor version to validate against.
```
updated.

suhsteve commented on src/scala/microsoft-spark-2-3/src/main/scala/org/apache/spark/util/dotnet/Utils.scala:
```diff
@@ -191,6 +193,45 @@ object Utils extends Logging {
+      logWarning(
+        s"Ignoring spark patch version. Spark version used: $sparkVersion. " +
+          s"Normalized spark version used: $normalizedSparkVersion. " +
+          s"Spark major.minor prefix used: $sparkMajorMinorVersionPrefix")
```
Removed trailing `.` from var definition.

imback82 commented on src/scala/microsoft-spark-2-3/src/main/scala/org/apache/spark/util/dotnet/Utils.scala:
```diff
@@ -191,6 +193,45 @@ object Utils extends Logging {
+
+    if (!normalizedSparkVersion.startsWith(s"$supportedSparkMajorMinorVersionPrefix.")) {
+      throw new IllegalArgumentException(
+        s"Unsupported spark version used: $sparkVersion. " +
+          s"Normalized spark version used: $normalizedSparkVersion. " +
+          s"Supported spark major.minor version: $supportedSparkMajorMinorVersionPrefix")
```
Let's quote the version string so that there is no confusion with dots in the version.
```suggestion
        s"Unsupported spark version used: '$sparkVersion'. " +
          s"Normalized spark version used: '$normalizedSparkVersion'. " +
          s"Supported spark major.minor version: '$supportedSparkMajorMinorVersionPrefix'")
```

suhsteve replied:
Added quotes.
